src/qibo/models/error_mitigation.py

```diff
@@ -51,51 +51,77 @@ def get_gammas(noise_levels, analytical: bool = True):
     """
-    if insertion_gate not in ("CNOT", "RX"):  # pragma: no cover
-        raise_error(
-            ValueError,
-            "Invalid insertion gate specification. Please select between 'CNOT' and 'RX'.",
-        )
-    if insertion_gate == "CNOT" and circuit.nqubits < 2:  # pragma: no cover
-        raise_error(
-            ValueError,
-            "Provide a circuit with at least 2 qubits when using the 'CNOT' insertion gate. "
-            + "Alternatively, try with the 'RX' insertion gate instead.",
-        )
 
-    i_gate = gates.CNOT if insertion_gate == "CNOT" else gates.RX
+    if global_unitary_folding:  # pragma: no cover
+        circuit_no_meas = circuit.__class__(**circuit.init_kwargs)
+        circuit_meas = circuit.__class__(**circuit.init_kwargs)
+        for gate in circuit.queue:
+            if gate.name != "measure":
+                circuit_no_meas.add(gate)
+            else:
+                circuit_meas.add(gate)
+
+        noisy_circuit = circuit.__class__(**circuit.init_kwargs)
+        noisy_circuit = circuit_no_meas
+        for _ in range(num_insertions):
+            noisy_circuit += circuit_no_meas.invert() + circuit_no_meas
+
+        noisy_circuit += circuit_meas
+
+    else:  # pragma: no cover
+        if insertion_gate is None or insertion_gate not in (
+            "CNOT",
+            "RX",
+        ):  # pragma: no cover
+            raise_error(
+                ValueError,
+                "Invalid insertion gate specification. Please select between 'CNOT' and 'RX'.",
+            )
+        if insertion_gate == "CNOT" and circuit.nqubits < 2:  # pragma: no cover
+            raise_error(
+                ValueError,
+                "Provide a circuit with at least 2 qubits when using the 'CNOT' insertion gate. "
+                + "Alternatively, try with the 'RX' insertion gate instead.",
+            )
 
-    theta = np.pi / 2
-    noisy_circuit = circuit.__class__(**circuit.init_kwargs)
+        i_gate = gates.CNOT if insertion_gate == "CNOT" else gates.RX
 
-    for gate in circuit.queue:
-        noisy_circuit.add(gate)
+        theta = np.pi / 2
+        noisy_circuit = circuit.__class__(**circuit.init_kwargs)
 
-        if isinstance(gate, i_gate):
-            if insertion_gate == "CNOT":
-                control = gate.control_qubits[0]
-                target = gate.target_qubits[0]
-                for _ in range(num_insertions):
-                    noisy_circuit.add(gates.CNOT(control, target))
-                    noisy_circuit.add(gates.CNOT(control, target))
-            elif gate.init_kwargs["theta"] == theta:
-                qubit = gate.qubits[0]
-                for _ in range(num_insertions):
-                    noisy_circuit.add(gates.RX(qubit, theta=theta))
-                    noisy_circuit.add(gates.RX(qubit, theta=-theta))
+        for gate in circuit.queue:
+            noisy_circuit.add(gate)
+
+            if isinstance(gate, i_gate):
```

Comment:
Why are you applying $U$ and $U^{\dagger}$ only when $U$ is found?
Is this a common practice? Otherwise, we could be more general and ask the user to select the position in the queue (or in the graph, speaking more `qibo-core` friendly) where to apply the gate-inverse mechanism.

Reply:
Maybe @AlejandroSopena do you have a clearer picture of the ZNE implementation?

Reply:
@MatteoRobbiati , what you propose can be implemented. In fact, it is common to fold some gates at random. However, if the fidelity of each gate is different, then the expected value in the zero-noise limit cannot be calculated as a linear superposition of expected values at different noise levels as we do. We would need to be implement  a numerical fit (see the comment in my review).